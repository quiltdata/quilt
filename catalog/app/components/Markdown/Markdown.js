import hljs from 'highlight.js';
import flow from 'lodash/flow';
import memoize from 'lodash/memoize';
import PT from 'prop-types';
import React from 'react';
import { setPropTypes } from 'recompose';
import Remarkable from 'remarkable';
import { replaceEntities, escapeHtml } from 'remarkable/lib/common/utils';
import styled from 'styled-components';

import { composeComponent } from 'utils/reactTools';


const highlight = (str, lang) => {
  if (lang === 'none') {
    return '';
  } else if (hljs.getLanguage(lang)) {
    try {
      return hljs.highlight(lang, str).value;
    } catch (err) {
      // istanbul ignore next
      console.error(err); // eslint-disable-line no-console
    }
  } else {
    try {
      return hljs.highlightAuto(str).value;
    } catch (err) {
      // istanbul ignore next
      console.error(err); // eslint-disable-line no-console
    }
  }
  // istanbul ignore next
  return ''; // use external default escaping
};

const escape = flow(replaceEntities, escapeHtml);

/**
 * Plugin for remarkable that disables image rendering.
 *
 * @param {Object} md Remarkable instance
 */
const disableImg = (md) => {
  // eslint-disable-next-line no-param-reassign
  md.renderer.rules.image = (tokens, idx) => {
    const t = tokens[idx];
    const src = escape(t.src);
    const title = t.title ? ` "${escape(t.title)}"` : '';
    const alt = t.alt ? escape(t.alt) : '';
    return `<span>![${alt}](${src}${title})</span>`;
  };
};

/**
 * Plugin for remarkable that adjusts link rendering:
 *
 *   - adds rel="nofollow" attribute
 *
 * @param {Object} md Remarkable instance
 */
const adjustLinks = (md) => {
  // eslint-disable-next-line no-param-reassign
  md.renderer.rules.link_open = (tokens, idx) => {
    const t = tokens[idx];
    const title = t.title ? ` title="${escape(t.title)}"` : '';
    return `<a href="${escapeHtml(t.href)}" rel="nofollow"${title}>`;
  };
};

/**
 * Get Remarkable instance based on the given options (memoized).
 *
 * @param {Object} options
 *
 * @param {boolean} images
 *   Whether to render images notated as `![alt](src title)` or skip them.
 *
 * @returns {Object} Remarakable instance
 */
const getRenderer = memoize(({ images }) => {
  const md = new Remarkable('full', {
    highlight,
    html: false,
    linkify: true,
    typographer: true,
  });
  md.use(adjustLinks);
  if (!images) md.use(disableImg);
  return md;
});

// Ensure that markdown styles are smaller than page h1, h2, etc. since
// they should appear as subordinate to the page's h1, h2
const Style = styled.div`
  display: block;
  overflow: auto;

  h1 code {
    background-color: inherit;
  }

  /* prevent horizontal overflow */
  img {
    max-width: 100%;
  }
`;

export default composeComponent('Markdown',
  setPropTypes({
    data: PT.string,
    className: PT.string,
    images: PT.bool,
  }),
  ({ data, className = '', images = true }) => (
    <Style
      className={`markdown ${className}`}
      dangerouslySetInnerHTML={{
        // would prefer to render in a saga but md.render() fails when called
        // in a generator
        __html: getRenderer({ images }).render(data),
      }}
    />
  ));
