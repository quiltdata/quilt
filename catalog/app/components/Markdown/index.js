/* Markdown */
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import Remarkable from 'remarkable';
import hljs from 'highlight.js';


const highlight = (str, lang) => {
  if (lang === 'none') {
    return '';
  } else if (hljs.getLanguage(lang)) {
    try {
      return hljs.highlight(lang, str).value;
    } catch (err) {
      console.error(err); // eslint-disable-line no-console
    }
  } else {
    try {
      return hljs.highlightAuto(str).value;
    } catch (err) {
      console.error(err); // eslint-disable-line no-console
    }
  }
  return ''; // use external default escaping
};

// TODO review for script injection attacks; html set to false as prelim
const md = new Remarkable('full', {
  highlight,
  html: false,
  linkify: true,
  typographer: true,
});

function Markdown({ data }) {
  // would prefer to render in a saga but md.render() fails when called in
  // a generator
  const html = { __html: md.render(data) };
  return (
    <Style className="markdown" dangerouslySetInnerHTML={html} />
  );
}

Markdown.propTypes = {
  data: PropTypes.string,
};

/* Ensure that markdown styles are smaller than page h1, h2, etc. since
 * they should appear as subordinate to the page's h1, h2 */
const Style = styled.div`
  overflow: auto;
  h1 {
    font-size: 1.7em;
  }

  h2 {
    font-size: 1.5em;
  }

  h3 {
    font-size: 1.4em;
  }

  h4 {
    font-size: 1.3em;
  }

  h5 {
    font-size: 1.2em;
  }

  h6 {
    font-size: 1.1em;
  }

  h1 code {
    background-color: inherit;
  }
`;

export default Markdown;
