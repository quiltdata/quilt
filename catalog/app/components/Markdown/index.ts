export * from './Markdown'
export { default } from './Markdown'
