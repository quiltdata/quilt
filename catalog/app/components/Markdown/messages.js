/* Markdown Messages */
import { defineMessages } from 'react-intl';

export default defineMessages({
  error: {
    id: 'app.containers.Markdown.error',
    defaultMessage: 'Something went wrong fetching README.md',
  },
  none: {
    id: 'app.containers.Markdown.none',
    defaultMessage: 'No README.md provided',
  },
});
