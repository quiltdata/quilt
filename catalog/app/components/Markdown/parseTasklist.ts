import type { Remarkable } from 'remarkable'

export interface CheckboxContentToken extends Remarkable.ContentToken {
  checked: boolean
}

const isChar = (charIndex: number, str: string, charPosition: number) =>
  str.charCodeAt(charPosition) === charIndex
const isOpenBracket = isChar.bind(null, 91)
const isCloseBracket = isChar.bind(null, 93)
const isCross = (str: string, charPosition: number) =>
  isChar(120, str, charPosition) || isChar(88, str, charPosition)
const isSpace = isChar.bind(null, 32)

// TODO: Wait for https://github.com/jonschlinkert/remarkable/pull/401 and then remove
export default function parseTasklist(state: Remarkable.StateInline) {
  const charPosition = state.pos

  if (charPosition === state.posMax) return false
  if (!isOpenBracket(state.src, charPosition)) return false

  const nextCharPosition = charPosition + 1
  if (isCloseBracket(state.src, nextCharPosition)) {
    state.push({
      type: 'tasklist',
      checked: false,
      level: state.level,
    } as CheckboxContentToken)
    state.pos = nextCharPosition + 1
    return true
  }

  if (isCross(state.src, nextCharPosition) || isSpace(state.src, nextCharPosition)) {
    const lastCharPosition = nextCharPosition + 1
    if (!isCloseBracket(state.src, lastCharPosition)) return false

    state.push({
      type: 'tasklist',
      checked: isCross(state.src, nextCharPosition),
      level: state.level,
    } as CheckboxContentToken)
    state.pos = lastCharPosition + 1
    return true
  }

  return false
}
