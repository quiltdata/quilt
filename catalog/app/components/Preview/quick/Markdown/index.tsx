import * as React from 'react'

import Skeleton from './Skeleton'

const RenderLazy = React.lazy(() => import('./Render'))

export { default as Skeleton } from './Skeleton'

export function Render(props: Parameters<typeof RenderLazy>[0]) {
  return (
    <React.Suspense fallback={<Skeleton />}>
      <RenderLazy {...props} />
    </React.Suspense>
  )
}
