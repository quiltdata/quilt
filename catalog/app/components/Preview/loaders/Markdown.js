import { dirname, resolve } from 'path'

import * as R from 'ramda'
import * as React from 'react'

import { getRenderer } from 'components/Markdown'
import * as AWS from 'utils/AWS'
import AsyncResult from 'utils/AsyncResult'
import * as NamedRoutes from 'utils/NamedRoutes'
import * as Resource from 'utils/Resource'
import pipeThru from 'utils/pipeThru'
import useMemoEq from 'utils/useMemoEq'

import { PreviewData, PreviewError } from '../types'
import * as utils from './utils'

function useImgProcessor(handle) {
  const sign = AWS.Signer.useResourceSigner()
  return useMemoEq([sign, handle], () =>
    R.evolve({
      src: (src) =>
        sign({
          ptr: Resource.parse(src),
          ctx: { type: Resource.ContextType.MDImg(), handle },
        }),
    }),
  )
}

function useLinkProcessor(handle) {
  const { urls } = NamedRoutes.use()
  const sign = AWS.Signer.useResourceSigner()
  return useMemoEq([sign, urls, handle], () =>
    R.evolve({
      href: R.pipe(
        Resource.parse,
        Resource.Pointer.case({
          Path: (p) => {
            const hasSlash = p.endsWith('/')
            const resolved = resolve(dirname(handle.key), p).slice(1)
            const normalized = hasSlash ? `${resolved}/` : resolved
            return hasSlash
              ? urls.bucketDir(handle.bucket, normalized)
              : urls.bucketFile(handle.bucket, normalized)
          },
          _: (ptr) =>
            sign({
              ptr,
              ctx: { type: Resource.ContextType.MDLink(), handle },
            }),
        }),
      ),
    }),
  )
}

export const detect = utils.extIn(['.md', '.rmd'])

function MarkdownLoader({ gated, handle, children }) {
  const processImg = useImgProcessor(handle)
  const processLink = useLinkProcessor(handle)
  const data = utils.useObjectGetter(handle, { noAutoFetch: gated })
  const processed = utils.useProcessing(
    data.result,
    (r) => {
      const contents = r.Body.toString('utf-8')
      const rendered = getRenderer({ images: true, processImg, processLink })(contents)
      return PreviewData.Markdown({ rendered })
    },
    [processImg, processLink],
  )
  const handled = utils.useErrorHandling(processed, { handle, retry: data.fetch })
  const result =
    gated && AsyncResult.Init.is(handled)
      ? AsyncResult.Err(PreviewError.Gated({ handle, load: data.fetch }))
      : handled
  return children(result)
}

export const Loader = function GatedMarkdownLoader({ handle, children }) {
  const data = utils.useGate(handle)
  const handled = utils.useErrorHandling(data.result, { handle, retry: data.fetch })
  return pipeThru(handled)(
    AsyncResult.case({
      _: children,
      Ok: (gated) => <MarkdownLoader {...{ gated, handle, children }} />,
    }),
  )
}
