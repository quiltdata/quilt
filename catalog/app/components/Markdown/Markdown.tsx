import cx from 'classnames'
import createDOMPurify from 'dompurify'
import hljs from 'highlight.js'
import 'highlight.js/styles/default.css'
import memoize from 'lodash/memoize'
import * as React from 'react'
import * as Remarkable from 'remarkable'
import { linkify } from 'remarkable/linkify'
import * as M from '@material-ui/core'

import { linkStyle } from 'utils/StyledLink'

import parseTasklist, { CheckboxContentToken } from './parseTasklist'

/* Most of what's in the commonmark spec for HTML blocks;
 * minus troublesome/abusey/not-in-HTML5 tags: basefont, body, center, dialog,
 * dir, fieldset, form, frame, frameset, head, html, iframe, link, main, menu,
 * menuitem, meta, noframes,  optgroup, option, source (we don't support audio),
 * track (we don't support video).
 *
 * I opted not to include UI tags (opt, optgroup); ditto for base, body, head,
 * meta, title
 * which shouldn't be needed
 */
const SANITIZE_OPTS = {
  ALLOWED_TAGS: [
    'input',
    'a',
    'abbr',
    'address',
    'article',
    'aside',
    'b',
    'blockquote',
    'caption',
    'code',
    'col',
    'colgroup',
    'dd',
    'details',
    'del',
    'div',
    'dl',
    'dt',
    'em',
    'figure',
    'figcaption',
    'footer',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'header',
    'hr',
    'i',
    'img',
    'ins',
    'legend',
    'li',
    'mark',
    'nav',
    'ol',
    'p',
    'param',
    'pre',
    'section',
    'span',
    'strong',
    'sub',
    'summary',
    'sup',
    'table',
    'tbody',
    'td',
    'tfoot',
    'th',
    'thead',
    'tr',
    'ul',
  ],
  FORBID_TAGS: ['style', 'script'],
  FORBID_ATTR: ['style'],
}

// TODO: switch to pluggable react-aware renderer
// TODO: use react-router's Link component for local links
const highlight = (str: string, lang: string) => {
  if (lang === 'none') {
    return ''
  }
  if (hljs.getLanguage(lang)) {
    try {
      return hljs.highlight(str, { language: lang }).value
    } catch (err) {
      // istanbul ignore next
      console.error(err) // eslint-disable-line no-console
    }
  } else {
    try {
      return hljs.highlightAuto(str).value
    } catch (err) {
      // istanbul ignore next
      console.error(err) // eslint-disable-line no-console
    }
  }
  // istanbul ignore next
  return '' // use external default escaping
}

interface RemarkableWithUtils extends Remarkable.Remarkable {
  // NOTE: Remarkable.Remarkable doesn't export utils
  utils: {
    unescapeMd: (str: string) => string
  }
}

const { unescapeMd } = (Remarkable as unknown as RemarkableWithUtils).utils

const checkboxHandler = (md: Remarkable.Remarkable) => {
  md.inline.ruler.push('tasklist', parseTasklist, {})
  md.renderer.rules.tasklist = (tokens, idx) =>
    (tokens[idx] as CheckboxContentToken).checked ? '☑' : '☐'
}

type AttributeProcessor = (attr: string) => string

function handleImage(process: AttributeProcessor, element: Element): Element {
  const attributeValue = element.getAttribute('src')
  if (!attributeValue) return element
  const result = process(attributeValue)
  element.setAttribute('src', result)

  const alt = element.getAttribute('alt')
  if (alt) {
    element.setAttribute('alt', unescapeMd(alt))
  }

  return element
}

function handleLink(process: AttributeProcessor, element: HTMLElement): Element {
  const attributeValue = element.getAttribute('href')
  if (typeof attributeValue !== 'string') return element
  const result = process(attributeValue)
  element.setAttribute('href', result)

  const rel = element.getAttribute('rel')
  element.setAttribute('rel', rel ? `${rel} nofollow` : 'nofollow')

  return element
}

function htmlHandler(
  processLink?: AttributeProcessor,
  processImage?: AttributeProcessor,
) {
  return (currentNode: Element): Element => {
    const element = currentNode as HTMLElement
    const tagName = currentNode.tagName?.toUpperCase()
    if (processLink && tagName === 'A') return handleLink(processLink, element)
    if (processImage && tagName === 'IMG') return handleImage(processImage, element)
    return currentNode
  }
}

interface RendererArgs {
  processImg?: AttributeProcessor
  processLink?: AttributeProcessor
  win?: Window
}

export const getRenderer = memoize(
  ({ processImg, processLink, win = window }: RendererArgs) => {
    const md = new Remarkable.Remarkable('full', {
      highlight,
      html: true,
      typographer: true,
    }).use(linkify)
    md.use(checkboxHandler)
    const purify = createDOMPurify(win)
    purify.addHook('uponSanitizeElement', htmlHandler(processLink, processImg))
    return (data: string) => purify.sanitize(md.render(data), SANITIZE_OPTS)
  },
)

interface ContainerProps {
  children: string
  className?: string
}

const useContainerStyles = M.makeStyles({
  root: {
    overflow: 'auto',

    '& a': linkStyle,

    '& h1 code': {
      backgroundColor: 'inherit',
    },

    /* prevent horizontal overflow */
    '& img': {
      maxWidth: '100%',
    },

    '& table': {
      maxWidth: '100%',
      width: '100%',

      'th, td': {
        lineHeight: '1.5em',
        padding: '8px',
        textAlign: 'left',
      },

      '&, th, td': {
        border: '1px solid #ddd',
      },
    },
  },
})

export function Container({ className, children }: ContainerProps) {
  const classes = useContainerStyles()
  return (
    <div
      className={cx(className, classes.root)}
      // eslint-disable-next-line react/no-danger
      dangerouslySetInnerHTML={{ __html: children }}
    />
  )
}

interface MarkdownProps extends RendererArgs, Omit<ContainerProps, 'children'> {
  data?: string
}

export default function Markdown({
  data,
  processImg,
  processLink,
  ...props
}: MarkdownProps) {
  return (
    <Container {...props}>
      {getRenderer({ processImg, processLink })(data || '')}
    </Container>
  )
}
