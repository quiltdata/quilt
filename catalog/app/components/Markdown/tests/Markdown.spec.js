import { mount } from 'enzyme';
import React from 'react';

import Markdown from '..';


const render = (data, props) =>
  mount(<Markdown data={data} {...props} />).render();

const codeExpr = 'const v = 1';
const hlAuto =
`\`\`\`
${codeExpr}
\`\`\``;
const hlNone =
`\`\`\`none
${codeExpr}
\`\`\``;
const hlJS =
`\`\`\`javascript
${codeExpr}
\`\`\``;
const hlBad =
`\`\`\`javascrip
${codeExpr}
\`\`\``;

const imgs = [
  '![the image](http://test "the image title")',
  '![](http://test)',
];

const links = [
  '[sup](/url)',
  '[sup](/url "link title")',
];

describe('components/Markdown', () => {
  describe('syntax highlighting', () => {
    it('should highlight the recognized syntax properly', () => {
      const result = render(hlJS);
      expect(result.find('code.language-javascript')).toHaveLength(1);
      expect(result).toMatchSnapshot();
    });

    it('should not highlight if syntax set to "none"', () => {
      const result = render(hlNone);
      expect(result.find('code.language-none')).toHaveLength(1);
      expect(result.find('code.language-none').html()).toMatch(codeExpr);
      expect(result).toMatchSnapshot();
    });

    it('should auto-highlight when syntax is not specified or not recognized', () => {
      const resultAuto = render(hlAuto);
      const resultBad = render(hlBad);
      expect(resultAuto.find('code').html()).toBe(resultBad.find('code').html());
      expect(resultBad.find('code.language-javascrip')).toHaveLength(1);
      expect(resultAuto).toMatchSnapshot();
      expect(resultBad).toMatchSnapshot();
    });
  });

  describe('images', () => {
    it('should be rendered if enabled (default)', () => {
      const result = render(imgs[0]);
      const resultImg = result.find('img');
      expect(resultImg).toHaveLength(1);
      expect(resultImg.attr('src')).toBe('http://test');
      expect(resultImg.attr('alt')).toBe('the image');
      expect(resultImg.attr('title')).toBe('the image title');
      expect(result).toMatchSnapshot();
    });

    it('should not be rendered if disabled', () => {
      imgs.forEach((markup) => {
        const result = render(markup, { images: false });
        const resultSpan = result.find('span');
        expect(result.find('img')).toHaveLength(0);
        expect(resultSpan).toHaveLength(1);
        expect(resultSpan.text()).toBe(markup);
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('links', () => {
    it('should be rendered with rel="nofollow" attribute', () => {
      links.forEach((markup) => {
        const result = render(markup);
        const resultLink = result.find('a');
        expect(resultLink).toHaveLength(1);
        expect(resultLink.attr('rel')).toBe('nofollow');
        expect(result).toMatchSnapshot();
      });
    });
  });
});
