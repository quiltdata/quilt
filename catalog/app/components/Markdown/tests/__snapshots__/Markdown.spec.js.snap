// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Markdown images should be rendered if enabled (default) 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <p>
    <img
      alt="the image"
      src="http://test"
      title="the image title"
    />
  </p>
  

</div>
`;

exports[`components/Markdown images should not be rendered if disabled 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <p>
    <span>
      ![the image](http://test "the image title")
    </span>
  </p>
  

</div>
`;

exports[`components/Markdown images should not be rendered if disabled 2`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <p>
    <span>
      ![](http://test)
    </span>
  </p>
  

</div>
`;

exports[`components/Markdown links should be rendered with rel="nofollow" attribute 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <p>
    <a
      href="/url"
      rel="nofollow"
    >
      sup
    </a>
  </p>
  

</div>
`;

exports[`components/Markdown links should be rendered with rel="nofollow" attribute 2`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <p>
    <a
      href="/url"
      rel="nofollow"
      title="link title"
    >
      sup
    </a>
  </p>
  

</div>
`;

exports[`components/Markdown syntax highlighting should auto-highlight when syntax is not specified or not recognized 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <pre>
    <code>
      <span
        class="hljs-keyword"
      >
        const
      </span>
       v = 
      <span
        class="hljs-number"
      >
        1
      </span>
      

    </code>
  </pre>
  

</div>
`;

exports[`components/Markdown syntax highlighting should auto-highlight when syntax is not specified or not recognized 2`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <pre>
    <code
      class="language-javascrip"
    >
      <span
        class="hljs-keyword"
      >
        const
      </span>
       v = 
      <span
        class="hljs-number"
      >
        1
      </span>
      

    </code>
  </pre>
  

</div>
`;

exports[`components/Markdown syntax highlighting should highlight the recognized syntax properly 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <pre>
    <code
      class="language-javascript"
    >
      <span
        class="hljs-keyword"
      >
        const
      </span>
       v = 
      <span
        class="hljs-number"
      >
        1
      </span>
      

    </code>
  </pre>
  

</div>
`;

exports[`components/Markdown syntax highlighting should not highlight if syntax set to "none" 1`] = `
.c0 {
  display: block;
  overflow: auto;
}

.c0 h1 code {
  background-color: inherit;
}

.c0 img {
  max-width: 100%;
}

<div
  class="markdown  c0"
>
  <pre>
    <code
      class="language-none"
    >
      const v = 1

    </code>
  </pre>
  

</div>
`;
